File: Tennis/csharp/2ndTry/TennisKata/Tennis.cs
using System;

namespace TennisKata
{
    public class Tennis
    {
        private int winsPlayer1, winsPlayer2;
        private const int love = 0;
        private const int forty = 3;

        public string PrintScore()
        {
            if (OnePlayerHasMoreThan(forty) && OnPlayerHas(love))
                return LeadingPlayer() + " wins";

            if (OnePlayerHasMoreThan(forty) && APlayerLeadByTwoOrMoreBalls())
                return LeadingPlayer() + " wins";

            if (BothPlayersHas(forty) && ScoreIsEqual())
                return "deuce";

            if (BothPlayersHas(forty) && APlayerLeadByOneBall())
                return "advantage " + LeadingPlayer();

            if (BothPlayersHas(forty) && APlayerLeadByTwoOrMoreBalls())
                return LeadingPlayer() + " wins";

            return string.Format("{0} - {1}", Score(winsPlayer1), Score(winsPlayer2));
        }

        private bool OnePlayerHasMoreThan(int score)
        {
            return winsPlayer1 > score || winsPlayer2 > score;
        }

        private bool OnPlayerHas(int score)
        {
            return winsPlayer1 == score || winsPlayer2 == score;
        }

        private bool BothPlayersHas(int score)
        {
            return winsPlayer1 >= score && winsPlayer2 >= score;
        }

        private string LeadingPlayer()
        {
            return winsPlayer1 > winsPlayer2 ? "Player1" : "Player2";
        }

        private bool ScoreIsEqual()
        {
            return winsPlayer1 == winsPlayer2;
        }

        private bool APlayerLeadByOneBall()
        {
            return winsPlayer1 == winsPlayer2 + 1 || winsPlayer2 == winsPlayer1 + 1;
        }

        private bool APlayerLeadByTwoOrMoreBalls()
        {
            return winsPlayer1 >= winsPlayer2 + 2 || winsPlayer2 >= winsPlayer1 + 2;
        }

        private static string Score(int balls)
        {
            if (balls == 1) return "fifteen";
            if (balls == 2) return "thirty";
            if (balls == 3) return "forty";
            
            return "love";
        }

        public void Player1WinsTheBall()
        {
            winsPlayer1++;
        }

        public void Player2WinsTheBall()
        {
            winsPlayer2++;
        }
    }
}

File: Tennis/ruby/FirstTry/tennis.rb
require 'dsl'

class Tennis
  include Dsl
  
  def initialize
    @wins_player1 = 0
    @wins_player2 = 0
    @score = {
      :love => 0,
      :fifteen => 1,
      :thirty => 2,
      :forty => 3
    }
  end
    
  def score
    is deuce! if both players has :forty 
    is advantage! if both players has :forty and leading player leads by 1
    is win! if one player has more than :forty and one player has :love
    is win! if both players has :forty and leading player leads by 2
    is still playing if !deuce? and !advantage? and !win?
    return @result
  end
  
  def is score
    @result = score
  end
  
  def deuce!
    @result = "deuce"
  end
  
  def advantage!
    @result = "advantage " + leading_player
  end
  
  def win!
    @result = leading_player + " wins"
  end
  
  def still_playing *args
    @result = "#{print @wins_player1} - #{print @wins_player2}"
  end
    
  def deuce? 
    both_players_has :forty
  end
  
  def advantage?
    both_players_has :forty and leading_player_leads_by 1
  end
  
  def win?
    (one_player_has_more_than :forty and one_player_has :love) or
    (both_players_has :forty and leading_player_leads_by 2)
  end
  
  def both_players_has score
    @wins_player1 >= @score[score] and @wins_player2 >= @score[score]
  end
  
  def leading_player
    return "Player 1" if @wins_player1 > @wins_player2
    return "Player 2"
  end

  def leading_player_leads_by balls
    @wins_player1 == @wins_player2 + balls or @wins_player2 == @wins_player1 + balls
  end
  
  def one_player_has score
    @wins_player1 == @score[score] or @wins_player2 == @score[score]
  end
  
  def one_player_has_more_than balls
    @wins_player1 > @score[balls] or @wins_player2 > @score[balls]
  end
  
  def print balls_won
    if balls_won == 1 then return "fifteen" end
    if balls_won == 2 then return "thirty" end
    if balls_won == 3 then return "forty" end
    
    return "love"
  end
  
  def player1_wins balls
    @wins_player1 += balls
  end
  
  def player2_wins balls
    @wins_player2 += balls
  end
end